Test CORS

loicgasser opened the pull request:
With this PR it is now possible to load external resources in geoadmin without proxy, provided that:

1. It supports https
2. It supports CORS


[Test Link](https://mf-geoadmin3.int.bgdi.ch/gal_proxify/index.html)

procrastinatio commented:
189/78 KB
https://procrastinatio.github.io/kml/bln-style.kmz
https://procrastinatio.github.io/kml/bln-style.kml


1.6/0.2 MB
https://procrastinatio.github.io/kml/simple_005.kml
https://procrastinatio.github.io/kml/simple_005.kmz

4.3/1.4 MB
https://procrastinatio.github.io/kml/simple_00005a.kml
https://procrastinatio.github.io/kml/simple_00005a.kmz


12/4.0 MB
https://procrastinatio.github.io/kml/simple_000005a.kml
https://procrastinatio.github.io/kml/simple_000005a.kmz

loicgasser commented:
Great ! Thanks a lot !!!

procrastinatio commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,44 @@ goog.provide('ga_urlutils_service');
+        // Test using a head request if the remote resource enables CORS
+        this.isCorsEnabled = function(url) {
+          var deferred = $q.defer();
+          $http.head(url, { timeout: 2500 }).then(function(response) {
```
Maybe a bit too long 2.5 sec

loicgasser replied:
yes, I reduced it to 1.5 sec, but if network is bad on mobile, we have a small buffer.

procrastinatio commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,44 @@ goog.provide('ga_urlutils_service');
+        this.proxifyUrl = function(url) {
+          var deferred = $q.defer();
+          if (!this.isBlob(url) && this.isHttps(url) &&
+              !this.isAdminValid(url) && !/.*kmz$/.test(url)) {
```
We may also consider that all domains under our control ('.bgdi.ch' and '.geo.admin.ch') do not need to be checked and support HTTPS and CORS by default.

loicgasser replied:
yes but you will break the certificate. but technically yes

gjn replied:
That's what the isAdminValid does. So all *.geo.admin.ch addresses are bypassed anyway.

procrastinatio commented on src/components/importkml/ImportKmlDirective.js:
```diff
@@ -51,24 +51,26 @@ goog.require('ga_urlutils_service');
+              }).then(function(response) {
+                var data = response.data;
+                var fileSize = response.headers('content-length');
+                if (gaKml.isValidFileContent(data) &&
```
If filesize is bigger than 4 MB, we cannot use the proxy and we are doomed. Maybe issue a warning.

loicgasser replied:
I am waiting on a final decision and service deployment before changing the limit. But yes we may change it.

gjn commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,44 @@ goog.provide('ga_urlutils_service');
+          var deferred = $q.defer();
+          $http.head(url, { timeout: 1500 }).then(function(response) {
+            // If the head request responded successfully it means that CORS
+            // is enabled
```
Are you sure about that assumption? Don't we need to check the response headers?

I always assumed a correctly formed pre-flight OPTIONS request is needed. See http://stackoverflow.com/questions/12173990/how-can-you-debug-a-cors-request-with-curl (among other resources, CORS is a beast)

loicgasser replied:
yes using curl and being in the browser is 2 different things. The browser will block the resource and as a result the request will be blocked.
I'll provide more test cases.

loicgasser replied:
Now you re right we may have false positive in case the server doesn t allow head requests. Head requests do pre-flight, like get. http://stackoverflow.com/questions/22107694/why-does-a-cross-origin-head-request-need-a-preflight-check
Best would be to simulate the option request and check the content or simply do a first get and use the proxy on error.

loicgasser replied:
```
$ curl -I https://www.wandern.ch/ewalc/images/ewalk2/696/geodata/geodata_de.xml
HTTP/1.1 200 OK
Date: Fri, 10 Mar 2017 09:46:32 GMT
Server: Apache
Last-Modified: Mon, 30 Jan 2017 20:13:09 GMT
ETag: "4d8d-5475571c34058"
Accept-Ranges: bytes
Content-Length: 19853
Vary: Accept-Encoding
NZSpeedy: OFF
NZX: OK
Content-Type: application/xml
```

They don't support CORS. It's blocked by the browser as expected and we have to use ogcproxy.


https://www.gipfelbuch.ch/global/kml/gipfel/1198/show

Now we have an issue when requesting a resource that has been moved, like the one above (redirection over http).
So we still need to handle this case.

Again this needs to go through ogcproxy as expected. I don

loicgasser replied:
nop not an issue :D

procrastinatio replied:
HEAD on https://www.wandern.ch/ewalc/images/ewalk2/696/geodata/geodata_de.xml gives a 200, but the requests fails  in an ajax context (nor CORS). So the test is good, it has to go though the proxy

procrastinatio replied:
The alternative would be trying to load the resource, without any testing, and use the proxy only if it fails, effectively discouraging the use of the proxy.

gjn commented:
Please test with the big 3. (send in email this morning).

Can you create testlinks for them?

gjn commented:
Ok, I tested with the big 3. All 3 will still pass by the ogcproxy. So at least 93% of traffic will still pass via ogxproxy.

Anyhow, this is great for other resources that properly support CORS/https.

gjn commented:
@oterral I'd like your opinion on this.

oterral commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,44 @@ goog.provide('ga_urlutils_service');
+        // Test using a head request if the remote resource enables CORS
+        this.isCorsEnabled = function(url) {
+          var deferred = $q.defer();
+          $http.head(url, { timeout: 1500 }).then(function(response) {
```
you can return directly the promise. No need to handle results.

```
return $http.head(url, { timeout: 1500 });
```

loicgasser replied:
yep of course..

oterral commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,44 @@ goog.provide('ga_urlutils_service');
           return url;
         };
 
+        this.proxifyUrl = function(url) {
```
I think the content of this function is a duplication of `needsProxy`, you could move the `isCorsEnabled` stuff in the `needsProxy` function then, promisifying :) the `needsProxy` function you keep the `proxifyUrl` function more understandable, and you get rid of `proxifyUrlInstant`:

```
this.proxifyUrl = function(url) {
  var defer = $q.defer();
  this.needsProxy(url).then(function() {
      defer.resolve(gaGlobalOptions.ogcproxyUrl + encodeURIComponent(url));
  }, function() {
      defer.resolve(url);
  })
  return defer.promise;
};
```

loicgasser replied:
yes, though proxify instant is used to determine the url for Cesium WMS sources. I don't think we can set the sources of WMS Cesium layers in an asynchronous way.

oterral replied:
You're right it'sa problem. May be we could add a `sync` parameter to `proxifyUrl` and `needsProxy` functions which skip the HEAD request if true. I think it worth a try.

gjn replied:
What about the sync here?

loicgasser replied:
This is not a duplicate we have `&&` operator instead of `||` in order to test CORS, we must statisfy all these conditions.

I can make one function with a sync param but honeslty, I don't see the point.

oterral replied:
The point is to have one function for all which returns a url or a promise which returns a url.

loicgasser replied:
it's either we have one function and duplicated code, see: `deferred.resolve(this.proxifyUrlInstant(url));` line 86, either we keep it as is.

oterral replied:
it's ok keep it as it is

loicgasser commented:
FYI I am waiting for https://github.com/geoadmin/mf-geoadmin3/pull/3596 to be merged before adressing any comment since these PR change the code in the same locations.

gjn commented:
Something I don't understand. The gipfelbuch kml returns a 200 on the HEAD request in the browser, but it's still passed to the ogcproxy:

https://mf-geoadmin3.int.bgdi.ch/gal_proxy_integration/index.html?swisssearch=https://www.skitourenguru.ch/calc_data/kml/Switzerland_RiskSymbol.kml

Shouldn't it try to load it directyl if HEAD returns 200?

Also, the proxy doesn't seem to work currently.

loicgasser commented:
Head returns a 200 but fails in an ajax context so it s correct because they don t have cors.
Function has been deleted and changed name because marc is working on creating the minimal rights.

loicgasser commented:
No matter what is the status code if they don t have cors it will fail.

procrastinatio commented:
The file https://www.skitourenguru.ch/calc_data/kml/Switzerland_RiskSymbol.kml is loaded without the proxy, but lacking CORS headers, it won't display.

If you use a serious site with correct headers (https://www.procrastinatio.org/kml/Les_Bayards-02-02-2015.kml), it works flawlessly.

gjn commented:
Ok, I think I understand now. Confusing that HEAD request shows ok in the networking tab. Error can be seen in the console tab.

@procrastinatio https://www.skitourenguru.ch/calc_data/kml/Switzerland_RiskSymbol.kml  is loaded via proxy (at least that's what I see in the console)

loicgasser commented:
Updated
@oterral do you see any significant limitations to this approach with ngeo import?

oterral commented:
On the approach no. On the integration maybe, because the  [transformUrl](https://github.com/geoadmin/mf-geoadmin3/pull/3596/files#diff-2037d4809c5b41498dbbfd34d25e52ecR185) function doesn't accept promise, I guess. Probably something we need to modify in ngeo.

loicgasser commented:
We need to change:

https://github.com/camptocamp/ngeo/blob/master/src/modules/import/importonline.js#L134

procrastinatio commented:
Buggy KML http://www.nabodat.ch/images/docs/kml/BOKA_1_5000.kml

procrastinatio commented on src/components/importkml/ImportKmlDirective.js:
```diff
@@ -51,11 +51,9 @@ goog.require('ga_urlutils_service');
 
             // Angularjs doesn't handle onprogress event
-            gaUrlUtils.proxifyUrl($scope.fileUrl).then(function(url) {
+            gaUrlUtils.proxifyUrl($scope.fileUrl).then(function(url, headers) {
```
where it is defined. Now the `x-api_key`header is not sent

procrastinatio commented:
`zappa` is slowly killing me (https://github.com/geoadmin/service-proxy/issues/25)...

oterral commented on src/components/importkml/ImportKmlDirective.js:
```diff
@@ -51,24 +51,27 @@ goog.require('ga_urlutils_service');
-              } else {
+            gaUrlUtils.proxifyUrl($scope.fileUrl).then(function(opts) {
+              $http.get(opts.url, {
+                headers: opts.headers,
```
it seems the headers are always the same here. Why not putting it directly here instead of opts obj.

oterral replied:
Forget the comment. It makes sense.

oterral commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,54 @@ goog.provide('ga_urlutils_service');
+            deferred.resolve({
+              url: this.proxifyUrlInstant(url),
+              headers: { 'x-api-key': gaGlobalOptions.proxyApiKey }
+            });
```
the deferred should returns only the url

oterral replied:
Forget too. i Understand why.

oterral commented on src/components/UrlUtilsService.js:
```diff
@@ -48,13 +48,54 @@ goog.provide('ga_urlutils_service');
+          var protocol = parts[1];
+          var resource = parts[3];
+          return gaGlobalOptions.proxyUrl + protocol + '/' +
+              encodeURIComponent(resource);
```
Add a comment to explain why we are doing this.

oterral replied:
Please add a comment

oterral commented:
Something still bother me with this implementation. We should create a service `gaProxy` 

Instead of doing:

```
gaUrlUtils.proxifyUrl(layer.url).then
```

we should use something : ```  gaProxy.get(layer.url) .then```

but I can live without it.

procrastinatio commented:
This

    $ curl  -H "Referer: https://tuttu" -o /dev/null -s -w "%{http_code}\n" https://service-proxy.prod.bgdi.ch/https/www.procrastinatio.org/kml/bln-style.kml
    403

And 

    $ curl  -H "Referer: https://map.geo.admin.ch" -o /dev/null -s -w "%{http_code}\n" https://service-proxy.prod.bgdi.ch/https/www.procrastinatio.org/kml/bln-style.kml
    200

gjn commented:
This is a go for next weeks deploy. The big 3 have tested and are ok with the changes.

Is ogcproxy in production ready? export PROXY_URL=//service-proxy.prod.bgdi.ch

I think this branch needs a rebase.

loicgasser commented:
waiting for https://github.com/camptocamp/ngeo/pull/2532 and then I'll rebase. yes code is production ready.

loicgasser commented:
jenkins test

loicgasser commented:
rebased

loicgasser commented:
jenkins test

oterral commented on src/components/map/MapService.js:
```diff
@@ -779,7 +779,7 @@ goog.require('ga_urlutils_service');
             camera: scene.camera,
             canvas: scene.canvas,
-            proxy: new Cesium.DefaultProxy(gaGlobalOptions.ogcproxyUrl)
+            proxy: new Cesium.DefaultProxy(gaGlobalOptions.proxyUrl + 'http/')
```
why this `+ 'http'` ?

oterral replied:
why not https?

loicgasser replied:
That might be an issue in fact.
DefaultProxy expects a full url? Here we use http per default in the background.

loicgasser replied:
I am a little puzzled with this part of the code. this function is only used in getOlLayerById but we never have kml data sources in our layersConfig. Can you explain when this is used?

oterral commented on src/components/map/KmlService.js:
```diff
@@ -63,8 +63,16 @@ goog.require('ngeo.fileService');
+                          '<href>' + gaGlobalOptions.proxyUrl + 'http/')
+                 .replace(new RegExp('<href>' + gaGlobalOptions.proxyUrl +
+                                     'https://', 'g'),
+                          '<href>' + gaGlobalOptions.proxyUrl + 'https/');
```
Use a classic indentation it's not understandable like this.

loicgasser commented:
it works the same way as before (instead of <service>?url=<url>, we have <service>/<protocol>/<resource>, I'll add the comments.
I have some issues updating the submodule with the approach you took, I'll propose something else.

oterral commented:
so it's not the same :)

loicgasser commented:
In the current Jenkins build we:
```
error: Your local changes to the following files would be overwritten by checkout:
	src/modules/import/importlocal.js
	src/modules/import/importonline.js
	src/modules/import/wmsgetcap.js
	src/modules/import/wmsgetcapitem.js
Please, commit your changes or stash them before you can switch branches.
```
when updating a module.
simply removing src/ngeo on cleanall is enough to ensure a proper reset after a cleanall.
I also removeed the `git submodule update --init` in Jenkins build script, I guess this is an old artefact.

loicgasser commented:
This will also fix:
https://github.com/geoadmin/mf-geoadmin3/issues/3783
as I updated ngeo.
I re-rebased and added the getCesiumProxy as discussed, I think it is more correct with Cesium as well.

oterral commented:
There is an issue with osm wms:

[Before](https://map.geo.admin.ch/?lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege,WMS%7C%7CLANDSAT%201975%7C%7Chttp:%2F%2Frips-gdi.lubw.baden-wuerttemberg.de%2Farcgis%2Fservices%2Fwms%2FUIS_0100000017900001%2FMapServer%2FWMSServer%7C%7C0%7C%7C1.3.0%7C%7Ctrue&layers_visibility=false,false,false,false,true&layers_timestamp=18641231,,,,&lon=8.36178&lat=47.41519&elevation=10314&heading=360.000&pitch=-40.090)

[With new proxy](https://mf-geoadmin3.int.bgdi.ch/gal_proxify/index.html?lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bav.haltestellen-oev,ch.swisstopo.swisstlm3d-wanderwege,WMS%7C%7CLANDSAT%201975%7C%7Chttp:%2F%2Frips-gdi.lubw.baden-wuerttemberg.de%2Farcgis%2Fservices%2Fwms%2FUIS_0100000017900001%2FMapServer%2FWMSServer%7C%7C0%7C%7C1.3.0%7C%7Ctrue&layers_visibility=false,false,false,false,true&layers_timestamp=18641231,,,,&lon=8.36178&lat=47.41519&elevation=10314&heading=360.000&pitch=-40.090)

loicgasser commented:
@oterral it's fixed thanks for pointing that out!! This WMS is serving gzip images, which does not make much sense, but we have to take into account this exception.

oterral commented:
perfect for me this PR is good to go

gjn commented:
A milestone!

Thanks everyone involved, @loicgasser @danduk82 @procrastinatio @oterral .
